File: packages/zynth-web-browser/ios/WebBrowserModule.swift
import Foundation
import SafariServices
import UIKit
import ZynthKit

final class WebBrowserModule: NSObject, ZynthModule, SFSafariViewControllerDelegate {
  let name: String = "WebBrowser"

  var exportedMethods: [String] {
    ["openBrowserAsync", "dismissBrowser", "warmUpAsync", "coolDownAsync"]
  }

  private weak var runtime: ZynthRuntime?
  private weak var safariController: SFSafariViewController?
  private var pendingRequestId: String?

  init(runtime: ZynthRuntime) {
    self.runtime = runtime
    super.init()
  }

  func call(method: String, args: ZynthArgs) throws -> Any? {
    switch method {
    case "openBrowserAsync":
      return try openBrowserAsync(args: args)
    case "dismissBrowser":
      return dismissBrowser()
    case "warmUpAsync":
      return ["warmedUp": true]
    case "coolDownAsync":
      return ["cooledDown": true]
    default:
      throw ZynthModuleError.methodNotExported(module: name, method: method)
    }
  }

  private func openBrowserAsync(args: ZynthArgs) throws -> Any? {
    if pendingRequestId != nil {
      throw NSError(domain: "ZynthWebBrowser", code: 1, userInfo: [NSLocalizedDescriptionKey: "Another browser request is already active"])
    }

    let requestId = args.string("requestId", default: UUID().uuidString)
    let options = try args.dict("options")
    let rawUrl = options["url"] as? String ?? ""
    let url = try validateBrowserUrl(rawUrl)

    pendingRequestId = requestId
    DispatchQueue.main.async { [weak self] in
      guard let self else { return }
      guard let presenter = self.topViewController() else {
        self.emitResult(requestId: requestId, type: "error", errorCode: "E_NO_VIEW_CONTROLLER", errorMessage: "No active view controller")
        self.pendingRequestId = nil
        return
      }

      let browser = SFSafariViewController(url: url)
      browser.delegate = self
      self.safariController = browser
      presenter.present(browser, animated: true) {
        self.emitResult(requestId: requestId, type: "opened", url: url.absoluteString)
      }
    }

    return ["status": "pending"]
  }

  private func dismissBrowser() -> [String: Any] {
    guard let safariController else {
      return ["dismissed": false]
    }
    let requestId = pendingRequestId
    pendingRequestId = nil
    DispatchQueue.main.async {
      safariController.dismiss(animated: true)
    }
    if let requestId {
      emitResult(requestId: requestId, type: "dismiss")
    }
    return ["dismissed": true]
  }

  func safariViewControllerDidFinish(_ controller: SFSafariViewController) {
    guard let requestId = pendingRequestId else {
      return
    }
    pendingRequestId = nil
    safariController = nil
    emitResult(requestId: requestId, type: "cancel")
  }

  private func validateBrowserUrl(_ raw: String) throws -> URL {
    let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else {
      throw NSError(domain: "ZynthWebBrowser", code: 2, userInfo: [NSLocalizedDescriptionKey: "WebBrowser URL cannot be empty"])
    }
    guard trimmed.count <= 4096 else {
      throw NSError(domain: "ZynthWebBrowser", code: 3, userInfo: [NSLocalizedDescriptionKey: "WebBrowser URL exceeds maximum length"])
    }
    guard let url = URL(string: trimmed), let scheme = url.scheme?.lowercased(), scheme == "http" || scheme == "https", url.host != nil else {
      throw NSError(domain: "ZynthWebBrowser", code: 4, userInfo: [NSLocalizedDescriptionKey: "WebBrowser URL must be valid http/https"])
    }
    return url
  }

  private func emitResult(
    requestId: String,
    type: String,
    url: String? = nil,
    errorCode: String? = nil,
    errorMessage: String? = nil
  ) {
    let payload: [String: Any] = [
      "requestId": requestId,
      "type": type,
      "url": url ?? NSNull(),
      "errorCode": errorCode ?? NSNull(),
      "errorMessage": errorMessage ?? NSNull(),
    ]

    runtime?.emitEvent(
      name: "WebBrowser.result",
      payload: payload
    )
  }

  private func topViewController(base: UIViewController? = nil) -> UIViewController? {
    let start: UIViewController?
    if let base {
      start = base
    } else {
      start = UIApplication.shared.connectedScenes
        .compactMap { $0 as? UIWindowScene }
        .flatMap(\.windows)
        .first(where: { $0.isKeyWindow })?
        .rootViewController
    }

    if let nav = start as? UINavigationController {
      return topViewController(base: nav.visibleViewController)
    }
    if let tab = start as? UITabBarController {
      return topViewController(base: tab.selectedViewController)
    }
    if let presented = start?.presentedViewController {
      return topViewController(base: presented)
    }
    return start
  }
}

File: packages/zynth-web-browser/android/ZynthWebBrowser/src/main/java/dev/zynth/webbrowser/WebBrowserModule.kt
package dev.zynth.webbrowser

import android.app.Activity
import android.content.ActivityNotFoundException
import android.graphics.Color
import androidx.browser.customtabs.CustomTabColorSchemeParams
import androidx.browser.customtabs.CustomTabsIntent
import com.zynth.kit.runtime.ZynthArgs
import com.zynth.kit.runtime.ZynthModule
import com.zynth.kit.runtime.ZynthRuntime
import org.json.JSONObject
import java.net.URI

internal class WebBrowserModule(
  private val activity: Activity,
  private val runtime: ZynthRuntime,
) : ZynthModule {
  override val name: String = "WebBrowser"
  override val exportedMethods: List<String> = listOf(
    "openBrowserAsync",
    "dismissBrowser",
    "warmUpAsync",
    "coolDownAsync",
  )

  override fun call(method: String, args: ZynthArgs): JSONObject {
    return when (method) {
      "openBrowserAsync" -> openBrowserAsync(args)
      "dismissBrowser" -> JSONObject().put("dismissed", false)
      "warmUpAsync" -> JSONObject().put("warmedUp", true)
      "coolDownAsync" -> JSONObject().put("cooledDown", true)
      else -> throw IllegalArgumentException("Unsupported method: $method")
    }
  }

  private fun openBrowserAsync(args: ZynthArgs): JSONObject {
    val requestId = args.getString("requestId")
    val options = args.getMap("options")
    val url = readString(options, "url")
    val uri = parseHttpUri(url)

    val builder = CustomTabsIntent.Builder()
    readString(options, "toolbarColor")?.let { color ->
      runCatching {
        builder.setDefaultColorSchemeParams(
          CustomTabColorSchemeParams.Builder()
            .setToolbarColor(Color.parseColor(color))
            .build()
        )
      }
    }

    val customTabsIntent = builder.build().apply {
      intent.addFlags(if (readBoolean(options, "createTask", false)) android.content.Intent.FLAG_ACTIVITY_NEW_TASK else 0)
    }

    return try {
      customTabsIntent.launchUrl(activity, android.net.Uri.parse(uri.toString()))
      emitResult(
        requestId = requestId,
        type = "opened",
        url = uri.toString(),
      )
      JSONObject().put("status", "opened")
    } catch (_: ActivityNotFoundException) {
      emitResult(
        requestId = requestId,
        type = "error",
        errorCode = "E_ACTIVITY_NOT_FOUND",
        errorMessage = "No browser activity found to handle Custom Tabs intent",
      )
      JSONObject().put("status", "error")
    }
  }

  private fun parseHttpUri(raw: String?): URI {
    val value = raw?.trim().orEmpty()
    if (value.isEmpty()) {
      throw IllegalArgumentException("WebBrowser URL cannot be empty")
    }
    if (value.length > 4096) {
      throw IllegalArgumentException("WebBrowser URL exceeds maximum length")
    }
    val uri = URI(value)
    val scheme = uri.scheme?.lowercase()
    if (scheme != "http" && scheme != "https") {
      throw IllegalArgumentException("WebBrowser URL must use http or https")
    }
    if (uri.host.isNullOrBlank()) {
      throw IllegalArgumentException("WebBrowser URL must include a host")
    }
    return uri
  }

  private fun readString(map: Map<String, Any?>, key: String): String? {
    val value = map[key] ?: return null
    return value as? String
  }

  private fun readBoolean(map: Map<String, Any?>, key: String, fallback: Boolean): Boolean {
    val value = map[key] ?: return fallback
    return value as? Boolean ?: fallback
  }

  private fun emitResult(
    requestId: String,
    type: String,
    url: String? = null,
    errorCode: String? = null,
    errorMessage: String? = null,
  ) {
    val payload = JSONObject()
      .put("requestId", requestId)
      .put("type", type)
      .put("url", url ?: JSONObject.NULL)
      .put("errorCode", errorCode ?: JSONObject.NULL)
      .put("errorMessage", errorMessage ?: JSONObject.NULL)
    runtime.emitEvent("WebBrowser.result", payload)
  }
}
